### PR TITLE
feat: Add Fashion Recommendation System using Deep Learning and Fashion-MNIST (#1799)

### LINKED ISSUE #1799: Fashion Recommendation System using Deep Learning and Fashion-MNIST
### Project Overview
I would like to add a new deep learning project based on the Fashion-MNIST dataset that not only classifies clothing items but also recommends visually similar fashion products using feature extraction techniques.

### Proposed Features
- Fashion item classification using CNN
- Feature extraction from trained deep learning model
- Similar outfit/fashion item recommendation
- Visualization of recommended clothing items
- Accuracy and loss performance graphs
- Clean and beginner-friendly notebook structure

### Tech Stack
- Python
- TensorFlow / Keras
- NumPy
- Matplotlib
- Scikit-learn

### Benefits
- Introduces contributors to practical deep learning applications
- Combines image classification with recommendation systems
- Improves learning experience for beginners in computer vision
- Adds a real-world fashion AI use case to the repository

### Additional Notes
I would like to work on this project under GSSoC'26. Kindly assign this issue to me.

### PR DESCRIPTION
**Pull request by Anijesh**
## Related Issue

Closes: #1799

- [x] Contributor (GSSoC)

---

### ✨ What does this PR do?

Adds **`Fashion_MNIST_Recommendation_System/`** — a content-based fashion recommender built on
Fashion-MNIST. Give it a clothing image and it returns the visually most similar items from a
catalogue.

The core idea: **the recommendations come for free from a classifier.** A CNN is trained to
tell the ten Fashion-MNIST categories apart, then its output layer is discarded and the
128-dimensional vector learned just underneath is kept. Images the network considers similar
land close together in that space, so nearest-neighbour search over those vectors *is* a
recommender — no second model, and no hand-labelled "these two items are similar" pairs.

It also answers the question most recommender demos skip: **does it actually work?** Retrieval
is scored with precision@k against a raw-pixel baseline.

### 🚀 Features implemented (all requested in the issue)

* **Fashion item classification using CNN** — two conv blocks with batch norm + dropout.
* **Feature extraction from the trained model** — the `embedding` layer reused as a 128-D
  feature extractor (6.1× compression from 784 pixels).
* **Similar item recommendation** — cosine nearest-neighbour index over a 20,000-item catalogue.
* **Visualization of recommended clothing items** — query-vs-recommendations grid, colour-coded
  by whether the retrieved item matches the query's category.
* **Accuracy and loss performance graphs** — plus a confusion matrix and a PCA view of the
  learned embedding space.
* **Clean, beginner-friendly notebook structure** — 8 numbered sections, every step explained.

---

### Classification

| Metric | Value |
|---|---|
| **Test accuracy** | **92.73 %** |
| Test loss | 0.2093 |
| Macro F1 | 0.928 |
| Training | 15 epochs, 942 s on CPU |

Per-class F1 runs from **0.990 (trouser)** to **0.783 (shirt)** — confusion sits exactly where a
human would hesitate: shirt vs t-shirt/top vs coat vs pullover.

### Recommendation quality — precision@k vs a raw-pixel baseline

Precision@k = share of recommended items in the same category as the query.
Random retrieval over ten balanced classes ≈ 0.10.

| k | CNN embeddings | Raw-pixel baseline | Improvement |
|:--:|:--:|:--:|:--:|
| 1 | **0.917** | 0.834 | **+0.083** |
| 5 | **0.918** | 0.805 | **+0.113** |
| 10 | **0.916** | 0.792 | **+0.124** |

The learned features win at every `k`, and the gap *widens* as `k` grows — the CNN has learned
what makes two garments the same kind of thing, not which images share bright pixels in the
same places.

Precision@5 by category: Sandal 0.989 · Bag 0.985 · Trouser 0.981 · Sneaker 0.977 ·
Ankle boot 0.931 · T-shirt/top 0.893 · Pullover 0.888 · Coat 0.877 · Dress 0.867 · **Shirt 0.810**.

---

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Code style update (formatting, local variables)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

---

## How Has This Been Tested?

**1. The notebook is committed executed** — 15 code cells, all with outputs, **10 embedded
figures, 0 errors**. Every number quoted above is from that committed run, not an estimate.

**2. Unit tests — 16 passed:**

```text
$ python -m unittest discover -s tests
Ran 16 tests in 6.9s

OK
```

They cover the retrieval and metric layer with synthetic embeddings, so they need **neither
TensorFlow nor the dataset** (TensorFlow is imported lazily, only inside the model-building
helpers). They assert that recommendations come from the query's cluster, that similarities are
sorted best-first, that `k` is capped at the gallery size, that a well-separated gallery scores
precision@k = 1.0 while random embeddings score near chance, and that misuse — querying before
`fit()`, mismatched label counts — raises instead of silently misbehaving.

**3. PEP 8** — `pycodestyle --max-line-length=100 fashion_recommender.py tests/` reports no issues.

**4. Reproducibility** — seeds are fixed with `keras.utils.set_random_seed(42)`; two independent
full runs produced identical accuracy and precision@k. The dataset ships with Keras, so there is
no manual download and no API key.

---

## Screenshots

**Recommendations — query in blue, its five nearest catalogue items to the right (green = same category)**

recommendations

It matches *style*, not merely category — the sandals share a strap silhouette, the dresses a
cut, the pullovers a printed panel.

**Training curves**

training curves

**Confusion matrix**

confusion matrix

**Learned embedding space (PCA)**

embedding space

**Dataset samples**

dataset samples

---

## Checklist

- [x] My code follows the guidelines of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly wherever it was hard to understand.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my feature works.
- [ ] Any dependent changes have been merged and published in downstream modules. *(n/a — no dependent changes)*

---

### Notes for reviewers

* **Not a duplicate of `Deep-Fashion-Recommendation-system`.** That project uses a *pretrained*
  VGG16 on a women's-fashion image set that must be downloaded from Google Drive, does no
  training, and evaluates by eye. This one trains a CNN from scratch on Fashion-MNIST, does
  classification **and** retrieval, measures precision@k against a baseline, and is fully
  reproducible with no download. The README has a side-by-side table.
* No existing files were modified except the root `README.md`, where the project is added to the
  Quick Project Index and the Deep Learning domain table.
* The root `requirements.txt` is **untouched** (workflow-tracked); the project ships its own.
* **Heads-up on the Notebook Health Check:** the workflow does not install TensorFlow, so it will
  log a warning for this notebook when it tries to execute it. The step is `continue-on-error`
  and emits warnings only — it affects every deep-learning notebook in the repo, not just this
  one. The committed outputs are from a real local run.
* **Possible trivial conflict with my other open PR #2223**, which also appends an entry to the
  same Quick Project Index row. Whichever merges second, I am happy to rebase and renumber.
* Suggested labels: `type:feature`, `level:intermediate`, `gssoc:approved`.

### CODE REVIEW

**Comment by Anijesh** [comment]
@Sibam-Paul requesting review 🙏 — this closes #1799.

Since I am contributing from a fork I cannot apply labels myself. Could you please add `type:feature`, `level:intermediate` and `gssoc:approved`?

Everything in the PR description comes from a real committed run — the notebook is checked in **executed, with all 10 figures and 0 errors**, and the numbers (92.73% test accuracy, precision@5 of 0.918 vs 0.805 for a raw-pixel baseline) are reproducible: seeds are fixed and two independent full runs gave identical results.

Two things worth flagging up front:
1. This is **not** a duplicate of `Deep-Fashion-Recommendation-system` — that one uses pretrained VGG16 on a Drive-hosted image set with no training and visual-only evaluation; this trains a CNN from scratch on Fashion-MNIST and measures retrieval quality against a baseline. Side-by-side comparison is in the project README.
2. The Notebook Health Check will warn on this notebook because the workflow does not install TensorFlow. That step is `continue-on-error` and affects every DL notebook in the repo.

**Review by Sibam-Paul — CHANGES_REQUESTED** [review]
Add this project under the existing Fashion recommendation project folder 
with the method of implementation you are using @Anijesh

**Comment by Anijesh** [comment]
Thanks for the review @Sibam-Paul 🙏 — done, and pushed.

**What changed:** the project now lives at **`Deep-Fashion-Recommendation-system/CNN_Embeddings_FashionMNIST/`**, named after the method it implements — a CNN trained for classification whose penultimate layer is reused as a 128-D embedding for k-NN retrieval.

To make the methods easy to compare from the folder root, I also added an **implementation methods** table to `Deep-Fashion-Recommendation-system/README.md`:

| Method | Location | Approach | Dataset |
|---|---|---|---|
| Pretrained VGG16 features | `fashion_recommendation_system_using_image_features.ipynb` | Frozen VGG16 as an off-the-shelf extractor + cosine similarity, no training | Women's fashion images (Drive) |
| Trained CNN embeddings | `CNN_Embeddings_FashionMNIST/` | CNN trained to classify, penultimate layer reused as a 128-D embedding for k-NN — classification **and** retrieval from one model, scored with precision@k vs a raw-pixel baseline | Fashion-MNIST (ships with Keras) |

Git recorded every file as a pure rename, so the diff is just the move plus the README updates. **No code or results changed** — the notebook keeps its committed outputs (29 cells, 10 figures, 0 errors) and the 16 unit tests still pass after the move. I also updated the notebook's path fallback, the comparison section, and the root README index links so nothing points at the old location.

Recap of the results, for convenience: **92.73%** test accuracy, and precision@5 of **0.918** with the learned embeddings vs **0.805** for a raw-pixel baseline at the same k.

recommendations

Happy to flatten it into the parent folder instead of a sub-folder if you'd prefer that layout — just say the word.

**Comment by Anijesh** [comment]
@Sibam-Paul bro merge this
